### PR TITLE
Add JSON download of reviews and metadata for Review Quality Collector.

### PR DESCRIPTION
**Pull request by prechelt**
Download will include:
- hotcrp_version
- reviews
- papers (metadata only)
- reviewers' contacts (without passwords)
- reviewform

-------------------

Hi Eddie,

we talked about this in, um, October.
The patch has become reasonably neat, small, and non-specialized; other HotCRP users may find it useful, too.
It's taken so long because I wanted to get far enough on the RQC side to be reasonably sure this export actually does the job.
I'd be very happy if you'd deploy this new version 2.102 on hotcrp.com soon, because that is now the event my work will have to wait for.

I thank you for your cooperation so far.

  Lutz

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,5 +5,6 @@ extra/d3/d3-hotcrp.min.js
 extra/d3/node_modules
 filestore
 logs
+tmp
```
Comment (kohler) [inline .gitignore]: no

File: src/sa/sa_get_json.php [hunk 2]
```diff
@@ -55,5 +55,49 @@ function run(Contact $user, $qreq, $ssel) {
...
+    function allow(Contact $user) {
+        return $user->is_manager();
+    }
+    function list_actions(Contact $user, $qreq, PaperList $pl, &$actions) {
+        $actions[] = [3070, $this->subname, "Reviews", "Reviews (JSON)"];
+    }
+    function run(Contact $user, $qreq, $ssel) {
+        $conf = $user->conf;
+        $results = [];
+        $results["hotcrp_version"] = HOTCRP_VERSION;
```
Comment (kohler) [inline src/sa/sa_get_json.php]: It's important to filter the selection to only those papers that the user can in fact manage. `$user->can_administer($prow)`

File: src/sa/sa_get_json.php [hunk 3]
```diff
@@ -55,5 +55,49 @@ function run(Contact $user, $qreq, $ssel) {
...
+        //----- get reviews:
+        $reviews = $conf->qe_raw("select * from PaperReview where paperId in (" . implode(",", $ssel->selection()) . ")");
+        $results["reviews"] = $reviews->fetch_all(MYSQLI_ASSOC);
+        //----- get papers:
+        $papers = $conf->qe_raw("select * from Paper where paperId in (" . implode(",", $ssel->selection()) . ")");
+        $the_papers = $papers->fetch_all(MYSQLI_ASSOC);
```
Comment (kohler) [inline src/sa/sa_get_json.php]: You should be using PaperInfo::fetch() to fetch a PaperInfo object.

File: src/sa/sa_get_json.php [hunk 4]
```diff
@@ -55,5 +55,49 @@ function run(Contact $user, $qreq, $ssel) {
...
+        $nullify_sha = function($paper) { unset($paper["sha1"]); return $paper; };
+        $the_papers = array_map($nullify_sha, $the_papers);
```
Comment (kohler) [inline src/sa/sa_get_json.php]: See below. You should be using PaperInfo::paper_json().

File: src/sa/sa_get_json.php [hunk 5]
```diff
@@ -55,5 +55,49 @@ function run(Contact $user, $qreq, $ssel) {
...
+            $the_ids . "))";
+        $contacts = $conf->qe_raw($query);
+        $results["contacts"] = $contacts->fetch_all(MYSQLI_ASSOC);
+        //----- get review form descriptors:
+        $reviewform = $conf->qe_raw("select data from Settings where name = 'review_form'");
+        $results["reviewform"] = $reviewform->fetch_all(MYSQLI_ASSOC);
+        //----- export:
+        header("Content-Type: application/json");
+        #header("Content-Disposition: attachment; filename=" . "review-data.json");
+        echo json_encode($results, JSON_PRETTY_PRINT | JSON_PARTIAL_OUTPUT_ON_ERROR | JSON_HEX_QUOT) . "\n";
```
Comment (kohler) [inline src/sa/sa_get_json.php]: Hi Lutz, I'm not comfortable with this approach of just dumping all data from the database straight into the JSON. You listed in your first email the information you needed for RQC. But this dump contains far more than that, in fact it is indiscriminate.

HotCRP already has JSON formats for papers and for reviews. The JSON format for papers is defined by GetJson_SearchAction. The JSON format for reviews is defined by ReviewForm::unparse_review_json and ReviewForm::unparse_full_json (the review form JSON). I promise nothing about keeping the database formats consistent from version to version. The Paper JSON, and to some extent the paper review JSON, include information that should not be broadcast, such as the “author review capability.”

Comment (kohler) [inline src/sa/sa_get_json.php]: (The JSON formats for papers and reviews also contain more information than you need for RQC, but they at least are already defined and do not contain *all* database data.)

**Comment by prechelt** [comment]
Hi Eddie,

I fully understand your point of view. I am sorry for the issues: I was struggling badly to find any working solution at all (the code being terse as it is and me not exactly a PHP expert).
With your hints it has become a lot better now.

As for the "I don't promise to keep the DB as it is": I guess I'll have to take that risk, which doesn't appear to be tooooo high. We'll see how it plays out.
Thanks again for your help.

  Lutz

**Comment by prechelt** [comment]
Hi Eddie,
I see "kohler requested changes" and a link "See review" behind that, but nothing appears when I use that link. I have also not received any notification. 
Did you maybe submit an empty review? 
If not, do you have any idea what my problem is? A web search came up with nothing.

**Comment by prechelt** [comment]
OK, I think now I got it: The link refers to the _first_ review, the one my second and third commits have reacted to.  (I use a vertical screen; my viewport hardly moved.)
I'd be happy to receive feedback on those -- or a merge.

**Comment by kohler** [comment]
Hi Lutz, my review asked you to use ReviewForm::unparse_review_json, which you did not do. So I did not merge.

I checked in something, however. You should check it out on your install.

**Comment by prechelt** [comment]
Hi Eddie, OK, I converted the RQC side of the adapter to the new export format and my integration tests now run again against a local HotCRP server.  
I noticed that the current version is already running on HotCRP com. I started preparing the rqctest conference instance there, but that's not finished yet. I'll tell you when the quasi-real test against rqctest.hotcrp.com has worked as well.

**Comment by kohler** [comment]
Closing this!

**Pull request closed without merging**